### PR TITLE
DRAFT: Reject AOD events for which the relabeling of photon candidate…

### PR DESCRIPTION
**Pull request by sstiefel19**
…s fails (DATA & MC)

This change impacts only scenarios where the relabeling of AOD photon
candidates is performed within the AliV0Reader and NOT within the actual
AliAnalysisTask.

General info: After an event has passed the V0Reader's event selection,
the photon candidates are getting imported from the delta AOD file and
relabeled. This means that for every photon (skipping the candidate from now on)
from the delta AOD file its daughter tracks are tried to match with
two tracks from the set of all tracks within the event and relabeled on
success.

Before this change: If for a photon the relabeling fails (i.e. that
 the photon's daughter tracks can't be found within the event), then
its daughter tracks are set to negative and the photon will create
an entry under 'no tracks' in the
GammaConvV1_...->ConvCuts_...->IsPhotonSelected histogram and there
will be an error message. The event will not get rejected. The log
can get big since the error is shown for each photon where it fails.

With this change: If for a photon the relabeling fails, it will create
an entry in the same histogram as above. In addition, the quality of the
event will be set to 2 such that all AnalysisTasks which check the event
quality (I guess all) will reject this event. This will become visible
in the GammaConvV1_...->Cut Number...->...ESD Histograms->NEvents
histogram under 'Miss MC or inc. ev.' An error is shown at maximum once
per event (even less since it would be identical AliErrors)
Moreover, the filename of the current AOD file be added to the tree
of broken files. (Only for the GammaConvV1 task with this change)

The problem with failing relabeling was observed at GA_PbPb_AOD_MC
trainrun 1544, file /alice/sim/2019/LHC19h2a/297481/AOD/027/
also see: https://alice.its.cern.ch/jira/browse/ALIROOT-8304

### CODE REVIEW

**Comment by amaringarcia** [comment]
Hi,
this change for relabeling may need be be also ported to any other task using converted gammas. or? like convCalo and Dalitz, or Omega,.... It just came to my mind when seeing the list of at the committed files.
Ana

**Comment by sstiefel19** [comment]
Hi Ana,
that part, that adds the filenames to the tree with broken filenames, I guess we would want to port to other tasks too, yes. The rejection of events itself would already be enabled with this change for any other task which checks the eventQuality  from the V0reader (which I guess are all)

**Comment by sstiefel19** [comment]
Another issue would scenarios, where the relabeling happens inside the AnalysisTask and not within the V0reader. But I guess that should be addressed in an extra effort.

**Comment by amaringarcia** [comment]
If the removal of event happens at the V0Reader , it should be enough....

File: PWGGA/GammaConvBase/AliV0ReaderV1.cxx [hunk 1]
```diff
@@ -1262,14 +1263,20 @@ Bool_t AliV0ReaderV1::GetAODConversionGammas(){
...
+          if(fRelabelAODs) {
+            relabelingWorkedForAll &= RelabelAODPhotonCandidates(gamma);}
+          if(fConversionCuts->PhotonIsSelected(gamma,fInputEvent)){
+            new((*fConversionGammas)[fConversionGammas->GetEntriesFast()]) AliAODConversionPhoton(*gamma);}
         }
       }
+      if (!relabelingWorkedForAll){
+        AliError("For one ore more photon candidate the AOD daughters could not be found. The labels of those were set to -999999 and the event will get rejected.");
+        return kFALSE;
+      }
```
Comment (dmuhlhei) [inline PWGGA/GammaConvBase/AliV0ReaderV1.cxx:1279]: What if  "GetEntries()" in L 1276 returns 0 - meaning we have no V0 candidates? In your fix, this case will also be flagged with event type 2, which must not be the case in my opinion. Could you please check this? We need to be really careful here...

Comment (sstiefel19) [reply PWGGA/GammaConvBase/AliV0ReaderV1.cxx]: Good catch, thanks :) I agree, if we have no V0 candidates (or none of them is passing the cut) we do not necessarily want to flag the event with type 2

File: PWGGA/GammaConv/AliAnalysisTaskGammaConvV1.h [hunk 2]
```diff
@@ -408,7 +408,7 @@ class AliAnalysisTaskGammaConvV1 : public AliAnalysisTaskSE {
 
     AliAnalysisTaskGammaConvV1(const AliAnalysisTaskGammaConvV1&); // Prevent copy-construction
     AliAnalysisTaskGammaConvV1 &operator=(const AliAnalysisTaskGammaConvV1&); // Prevent assignment
-    ClassDef(AliAnalysisTaskGammaConvV1, 48);
+    ClassDef(AliAnalysisTaskGammaConvV1, 49);
```
Comment (sstiefel19) [inline PWGGA/GammaConv/AliAnalysisTaskGammaConvV1.h:411]: I just see this has increased in the meantime, so I guess I need to increase it once more?

**Comment by dmuhlhei** [comment]
Ok, I'm fine with the PR. 

@FriederikeBock and/or @nschmidtALICE  could you please have a look as well and approve this request? We should double check these changes

**Comment by nschmidtALICE** [comment]
PR also looks okay from my side.

**Comment by nschmidtALICE** [comment]
+test

**Comment by nschmidtALICE** [comment]
+1

**Pull request merged**